libnetwork/overlay: remove dead code

akerouanton opened the pull request:
**- What I did**

All the code related to Classic Swarm is removed from the overlay driver, as well as all the code paths becoming dead because of that.

**- How to verify it**

CI should be green. It can also be tested by forming a swarm cluster with two dev containers:

```
dev1$ docker network create --driver overlay --scope global --attachable testnet
dev1$ docker run -d --rm --name whoami --network testnet traefik/whoami
dev2$ docker run --rm -t --name netshoot --network testnet nicolaka/netshoot curl http://whoami
```

**- A picture of a cute animal (not mandatory but encouraged)**

![](https://lemagdesanimaux.ouest-france.fr/images/dossiers/2021-08/manchot-empereur-102100.jpg)

akerouanton committed:
libnetwork/netutils: remove dead code

- GetIfaceAddr is unused since moby/libnetwork@e51ead59
  (moby/libnetwork#670).
- ValidateAlias and ParseAlias are unused since moby/moby@0645eb84
  (moby/moby#42539).

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
libnetwork/netlabel: remove dead code

- LocalKVProvider, LocalKVProviderURL, LocalKVProviderConfig,
  GlobalKVProvider, GlobalKVProviderURL and GlobalKVProviderConfig
  are all unused since moby/libnetwork@be2b6962 (moby/libnetwork#908).
- GlobalKVClient is unused since 0fa873c and c8d2c6e.
- MakeKVProvider, MakeKVProviderURL and MakeKVProviderConfig are unused
  since 96cfb076 (moby/moby#44683).
- MakeKVClient is unused since 142b5229 (moby/moby#44875).

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton commented:
Failing job seems to be due to a flaky test.

I'm wondering if other `Global*` netlabels could be removed. I'll check that tomorrow.

akerouanton committed:
libnetwork: remove ovrouter cmd

This command was useful when overlay networks based on external KV store
was developed but is unused nowadays.

As the last reference to OverlayBindInterface and OverlayNeighborIP
netlabels are in the ovrouter cmd, they're removed too.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
libnetwork: remove global store from overlay driver

The overlay driver was creating a global store whenever
netlabel.GlobalKVClient was specified in its config argument. This
specific label is not used anymore since 142b522 (moby/moby#44875).

golangci-lint now detects dead code. This will be fixed in subsequent
commits.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
libnetwork: remove unused props from windows overlay driver

The overlay driver was creating a global store whenever
netlabel.GlobalKVClient was specified in its config argument. This
specific label is unused anymore since 142b522 (moby/moby#44875).

It was also creating a local store whenever netlabel.LocalKVClient was
specificed in its config argument. This store is unused since
moby/libnetwork@9e72136 (moby/libnetwork#1636).

Finally, the sync.Once properties are never used and thus can be
deleted.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
libnetwork/overlay: remove Serf-based clustering

Prior to 0fa873c, the serf-based event loop was started when a global
store was available. Since there's no more global store, this event loop
and all its associated code is dead.

Most dead code detected by golangci-lint in prior commits is now gone.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
libnetwork/overlay: remove host mode

Linux kernel prior to v3.16 was not supporting netns for vxlan
interfaces. As such, moby/libnetwork#821 introduced a "host mode" to the
overlay driver. The related kernel fix is available for rhel7 users
since v7.2.

This mode could be forced through the use of the env var
_OVERLAY_HOST_MODE. However this env var has never been documented and
is not referenced in any blog post, so there's little chance many people
rely on it. Moreover, this host mode is deemed as an implementation
details by maintainers. As such, we can consider it dead and we can
remove it without a prior deprecation warning.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

corhere commented on libnetwork/drivers/overlay/overlay.go:
```diff
@@ -184,10 +171,6 @@ func (d *driver) configure() error {
-		return nil
-	}
-
 	if d.vxlanIdm == nil {
 		return d.initializeVxlanIdm()
 	}
```
When `d.store == nil` the function would return early, before reaching this block. Therefore this block should be unreachable when constant-propagating `(d.store == nil) == true`.

```suggestion
```

akerouanton replied:
Oh and that explains why I couldn't get how/when `d.vxlanIdm` was used. Good news is, we can completely throw out VNI allocation from this driver :slightly_smiling_face:

corhere commented on libnetwork/drivers/overlay/overlay.go:
```diff
@@ -205,7 +188,7 @@ func (d *driver) initializeVxlanIdm() error {
 	}
 
-	d.vxlanIdm, err = idm.New(d.store, "vxlan-id", vxlanIDStart, vxlanIDEnd)
+	d.vxlanIdm, err = idm.New(nil, "vxlan-id", vxlanIDStart, vxlanIDEnd)
```
Woohoo, datastore support can be ripped out of package `idm`, which is the only user of `bitseq.Handle`... WE CAN DELETE PACKAGE `bitseq` SOON! 🎉 🎉

corhere commented on libnetwork/drivers/overlay/ov_network.go:
```diff
@@ -1083,35 +1023,7 @@ func (n *network) obtainVxlanID(s *subnet) error {
-		}
-		return nil
-	}
+	return fmt.Errorf("no valid vxlan id and no datastore configured, cannot obtain vxlan id")
```
What datastore? Never heard of him. Jokes aside, the error message could use an edit for a global-datastore-less world.

akerouanton replied:
Not sure if you're reviewing the whole diff or commits one by one. I didn't change this error message, instead I'm just getting rid of this method in a latter commit (I'm going to push it latter today) :slightly_smiling_face:

corhere commented on libnetwork/drivers/windows/overlay/overlay_windows.go:
```diff
@@ -23,7 +23,6 @@ const (
 	config     map[string]interface{}
 	networks   networkTable
-	store      datastore.DataStore
 	localStore datastore.DataStore
```
Somehow I'm not surprised that `store` wasn't actually used by the driver implementation. And just as unsurprisingly, neither is `localStore`. Mind cleaning that up too while you're at it?

akerouanton replied:
`once` and `joinOnce` too. I'm removing all of them :wink:

corhere reviewed (changes requested):
Almost everything LGTM. Has overlay host mode been documented as deprecated? We might not be able to rip it out unless it's gone through a deprecation cycle.

corhere commented:
Just discussed in today's maintainers call: host mode will need to go through a deprecation cycle. We can document it as deprecated now, deprecate in code for v-next, and remove in v-next+1. Mind splitting the last commit into a dedicated PR?

akerouanton committed:
libnetwork: remove local store from overlay driver

Since the previous commit, data from the local store are never read,
thus proving it was only used for Classic Swarm.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
libnetwork/overlay: remove vni allocation

VNI allocations made by the overlay driver were only used by Classic
Swarm. With Swarm v2 mode, the driver ovmanager is responsible of
allocating & releasing them.

Previously, vxlanIdm was initialized when a global store was available
but since 142b522, no global store can be instantiated. As such,
releaseVxlanID actually does actually nothing and iptables rules are
never removed.

The last line of dead code detected by golangci-lint is now gone.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
libnetwork/overlay: remove live-restore support

The overlay driver in Swarm v2 mode doesn't support live-restore, ie.
the daemon won't even start if the node is part of a Swarm cluster and
live-restore is enabled. This feature was only used by Swarm Classic.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
libnetwork/overlay: don't lock network when accessing subnet vni

Since a few commits, subnet's vni don't change during the lifetime of
the subnet struct, so there's no need to lock the network before
accessing it.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

corhere commented:
Since host mode is an implementation detail rather than a user-facing feature, and is only needed to support overlay networks on long-deprecated kernels, that code can be dropped without first going through a deprecation cycle. It'll need a changelog entry, though.

akerouanton committed:
libnetwork/overlay: remove KVObject implementation

Since 0fa873c, there's no function writing overlay networks to some
datastore. As such, overlay network struct doesn't need to implement
KVObject interface.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

corhere reviewed (changes requested):
Could we please save the new code and behaviour changes for a follow-up PR? Let's keep this PR on topic: removing dead code. Lazily cleaning up stale sandboxes may appear dumb, and it might be, but it also might have been done that way for a good reason. (Containers can survive a daemon restart, e.g. over an upgrade. Lazy clean-up might be a way to not disrupt the containers before Swarm has a chance to replace them with new ones.)

akerouanton commented on libnetwork/drivers/overlay/ov_network.go:
```diff
@@ -228,18 +232,6 @@ func (d *driver) DeleteNetwork(nid string) error {
-		return err
-	}
-
-	if n.secure {
-		for _, vni := range vnis {
-			programMangle(vni, false)
-			programInput(vni, false)
-		}
-	}
```
Yes, I'll submit a follow-up PR to properly clean-up those rules.

akerouanton commented:
> Could we please save the new code and behaviour changes for a follow-up PR? Let's keep this PR on topic: removing dead code.

Right, I'll move the incriminated commits to another branch for now.

> Lazily cleaning up stale sandboxes may appear dumb, and it might be, but it also might have been done that way for a good reason. (Containers can survive a daemon restart, e.g. over an upgrade. Lazy clean-up might be a way to not disrupt the containers before Swarm has a chance to replace them with new ones.)

I think it was meaningful when Classic Swarm was supported, but I don't see any reason to do that anymore with only Swarm Mode supported since it doesn't support live-restore:

1. dockerd won't start if it's part of a swarm cluster when live-restore is enabled (dockerd errors out with: `Error starting cluster component: --live-restore daemon configuration is incompatible with swarm mode`) ;
2. It returns an error if you try to join/init a cluster whereas live-restore is enabled (it fails with: `Error response from daemon: --live-restore daemon configuration is incompatible with swarm mode`) ;

Now that being said, I think that's not enough to prove my point. I don't want to spend time on this right now but I'll resubmit those changes if I come up with stronger arguments.

corhere reviewed (approved):
LGTM!

thaJeztah commented on libnetwork/netutils/utils.go:
```diff
@@ -62,35 +62,6 @@ func NetworkRange(network *net.IPNet) (net.IP, net.IP) {
 	return firstIP, lastIP
 }
 
-// GetIfaceAddr returns the first IPv4 address and slice of IPv6 addresses for the specified network interface
-func GetIfaceAddr(name string) (net.Addr, []net.Addr, error) {
-	iface, err := net.InterfaceByName(name)
-	if err != nil {
-		return nil, nil, err
-	}
-	addrs, err := iface.Addrs()
```
Just a random comment 🤔 

Haven't been through all changes yet, but I saw this being removed when I looked from my phone, and I was curious why it wasn't used (as some options accept either an IP-address or an interface name). But it looks like we have various "variations" on a a theme (some are test-utilities, so perhaps irrelevant). I haven't compared them fully (to see if there's specific differences between them), but was wondering if there would be use for a common implementation (or if it would be worth looking into);

- getBindAddr https://github.com/moby/moby/blob/dd3b71d17c614f837c4bba18baed9fa2cb31f1a4/libnetwork/agent.go#L58-L62
- resolveInterfaceAddr https://github.com/moby/moby/blob/dd3b71d17c614f837c4bba18baed9fa2cb31f1a4/daemon/cluster/listen_addr.go#L161-L168
- getExternalAddress https://github.com/moby/moby/blob/dd3b71d17c614f837c4bba18baed9fa2cb31f1a4/integration/container/nat_test.go#L121-L126
- setupDriver https://github.com/moby/moby/blob/dd3b71d17c614f837c4bba18baed9fa2cb31f1a4/libnetwork/drivers/overlay/overlay_test.go#L66-L70

corhere replied:
`libnetwork.getBindAddr` returns an arbitrary interface IP address which is not an IPv6 link-local address, with no preference of IP address family. On the other hand, `cluster.resolveInterfaceAddr` prefers IPv4 over IPv6, neglects to skip over IPv6 link-local addresses, and returns an error if there is more than one address of either family associated with the interface. Since IPv6 requires a link-local address be assigned to each interface in addition to any routable addresses, `cluster.resolveInterfaceAddr` is broken for any interface which has a routable IPv6 address associated.

I question the utility of both functions. While it is kinda convenient to let the administrator specify an interface name instead of an address, specifying an interface implies a promise which the daemon can't currently keep: that the daemon will dynamically handle changes to the IP addresses associated with the interface.

thaJeztah replied:
>  question the utility of both functions. While it is kinda convenient to let the administrator specify an interface name instead of an address, specifying an interface implies a promise which the daemon can't currently keep: that the daemon will dynamically handle changes to the IP addresses associated with the interface.

Yeah, swarm doesn't handle IP changes at all. The "allow interface" was mostly out of convenience (ISTR part of that was created at the time for automated setups of clusters)

It still feels like a good thing to verify the various code paths to check if they're consistent (and do the right thing).

Definitely not for this PR though!

thaJeztah reviewed (approved):
LGTM!
